virttest/qemu_devices/qdevices.py

```diff
@@ -1143,33 +1143,55 @@ def _cmdline_json(self):
         command_dict = {}
         out = "-%s " % self.type
 
-        usb_storage_driver = self.params.get("driver") == "usb-storage"
+        usb_driver = self.get_param("driver", "").startswith("usb-")
 
-        expect_string_val = ["write-cache", "disable-legacy"]
+        pcic = self.get_param("driver") in ("pcie-root-port")
+
+        pvpanic = self.get_param("driver") in ("pvpanic")
+
+        expect_string_val = ("write-cache", "disable-legacy", "intremap",
+                             "serial")
 
         for key, val in self.params.items():
             # wwn needs to be presented as hexadecimal
-            if key in ["wwn"]:
+            # port from device ( "driver": "pcie-root-port" )
+            if key in ("wwn") or (key == "port" and pcic) \
+                    or (key == "ioport" and pvpanic):
                 command_dict[key] = int(val, 16)
             # physical_block_size from device ("driver": "scsi-hd")
             # logical_block_size from device ("driver": "scsi-hd")
             # bootindex from device ("driver": "scsi-hd")
             # max_sectors from device ("driver": "virtio-scsi-pci")
             # num_queues from device ("driver": "virtio-scsi-pci")
             # virtqueue_size from device ("driver": "virtio-scsi-pci")
-            elif key in ["physical_block_size", "logical_block_size",
+            # period, max-bytes from device ("driver": "virtio-rng-pci")
+            # max-write-zeroes-sectors, queue-size, max-discard-sectors,
+            # num-queues from device ("driver": "virtio-blk-pci")
+            # host_mtu, speed, vectors from
+            # device ( "driver": "virtio-net-pci" )
+            # node from device ("driver": "pc-dimm")
+            # events from device("driver": "pvpanic")
+            # min_io_size, opt_io_size from device ( "driver": "usb-storage" )
+            elif key in ("physical_block_size", "logical_block_size",
                          "bootindex", "max_sectors", "num_queues",
-                         "virtqueue_size"]:
+                         "virtqueue_size", "period", "max-bytes",
+                         "max-write-zeroes-sectors", "queue-size",
+                         "max-discard-sectors", "num-queues", "host_mtu",
+                         "speed", "vectors", "node", "events", "min_io_size",
+                         "opt_io_size"):
                 command_dict[key] = int(val)
-            # port from device ("driver": "usb-storage")
-            elif key == "port" and usb_storage_driver:
+            # port from usb related driver
+            elif key == "port" and usb_driver:
                 command_dict[key] = str(val)
+            elif val == "NO_EQUAL_STRING":
+                if usb_driver and key == "serial":
+                    command_dict[key] = "on"
```

Comment:
For `usb-storage`, I agree with this change, to set it to `on`, actually this is a qemu change that deprecates to leave it as empty and should be an improvement on the test case side.

But don't forget that the original design is only `key` in raw format,
https://github.com/avocado-framework/avocado-vt/blob/97f1cbaab5dd9e03a6a981d5adc1a2f70555a6f4/virttest/qemu_devices/qdevices.py#L444-L448
https://github.com/avocado-framework/avocado-vt/blob/97f1cbaab5dd9e03a6a981d5adc1a2f70555a6f4/virttest/qemu_vm.py#L517-L520
In order to convert this to JSON format, maybe better to parse it to `{"key": ""} by default. But currently we don't have such scenarios, just a kind reminder, it's up to you.

Reply:
> For `usb-storage`, I agree with this change, to set it to `on`, actually this is a qemu change that deprecates to leave it as empty and should be an improvement on the test case side.
> 
> But don't forget that the original design is only `key` in raw format,
> 
> https://github.com/avocado-framework/avocado-vt/blob/97f1cbaab5dd9e03a6a981d5adc1a2f70555a6f4/virttest/qemu_devices/qdevices.py#L444-L448
> 
> 
> https://github.com/avocado-framework/avocado-vt/blob/97f1cbaab5dd9e03a6a981d5adc1a2f70555a6f4/virttest/qemu_vm.py#L517-L520
> 
> 
> In order to convert this to JSON format, maybe better to parse it to `{"key": ""} by default. But currently we don't have such scenarios, just a kind reminder, it's up to you.

Thanks to @PaulYuuu response.